### PR TITLE
Update app layout based on CSS grid

### PR DESCRIPTION
**Pull request by zhuiks**
This solves #77
The idea discussed in #299

### Summary of changes:
* refactor main layout inside appContainer with CSS grid
* refactor layout inside verseListTab with CSS grid
* responsive tagList and dictionary sections inside sidePanel implemented in pure CSS (without a need to adjust sizes with javascript)
* on small screen sizes (< 1200px) sidePanel is hedden and dictionary section shows underneath BibleBrowser

### CODE REVIEW

**Comment by zhuiks** [comment]
`dictionary_info_box` can be converted to a custom element, but maybe I'll do it with another PR

**Comment by tobias-klein** [comment]
Thanks! I'll have a look at this on the weekend. Let's try to get #262 merged! There are a few review comments pending.

**Comment by zhuiks** [comment]
What else needs to be refactored to not listen on `window` `resize`

Currently, I see all methods are in `uiHelper`:
`resizeVerseList`, `adaptVerseList`, `resizeAppContainer` 

Instead of adding classes via Javascript is better to deal with it via CSS media queries

**Comment by tobias-klein** [comment]
I just played a bit with this as a user (I have not looked at the code changes yet).

Observations / feedback:
- We should keep the option of having the dictionary below the tag list - from a formatting perspective this makes more sense because the dictionary tends to need more vertical space than horizontal space, especially if you have multiple dictionaries installed.
- So I think the dictionary should be below the tag list in the side bar by default and if there is no sidebar then it can be shown below the Bible browser.
- When the left side bar is not shown there is currently some horizontal space left on the left of the Bible browser (too much).

**Comment by zhuiks** [comment]
Thanks for the feedback

>     * We should keep the option of having the dictionary below the tag list - from a formatting perspective this makes more sense because the dictionary tends to need more vertical space than horizontal space, especially if you have multiple dictionaries installed.
> 
>     * So I think the dictionary should be below the tag list in the side bar by default and if there is no sidebar then it can be shown below the Bible browser.

Let's discuss it a bit more in #299. I'll convert this to a draft PR after resolving conflicts


>     * When the left side bar is not shown there is currently some horizontal space left on the left of the Bible browser (too much).
That's fixable. It's the CSS grid gap. But probably is better to do gutter with margins. Then when one area is collapsed margings/padding would be gone also

**Comment by zhuiks** [comment]
Please see initial comment for summary of changes.

**Comment by zhuiks** [comment]
I found that this layout breaks on Android 10.0 and lower. I'll see what is the cause.

FYI: Grid is supported starting from Chrome/WebView 57. 
Maybe we should bump minimal supported Chrome version from 55 to 57.

**Comment by tobias-klein** [comment]
> FYI: Grid is supported starting from Chrome/WebView 57.
> Maybe we should bump minimal supported Chrome version from 55 to 57.

No problem. Feel free to make that adjustment in `app/frontend/platform_init.js`.

File: app/frontend/components/dictionary_info_box.js [hunk 1]
```diff
@@ -261,8 +260,11 @@ class DictionaryInfoBox {
     var blueLetterTranslations = ['KJV', 'NASB', 'ASV', 'WEB'];
     if (!blueLetterTranslations.includes(bible)) {
       bible = 'KJV';
+    } else if (bible === 'NASB') {
```
Comment (tobias-klein) [inline app/frontend/components/dictionary_info_box.js:263]: Thanks. This must be a recent change on BLB.

File: app/frontend/components/dictionary_info_box.js [hunk 2]
```diff
@@ -261,8 +260,11 @@ class DictionaryInfoBox {
...
+      bible = 'NASB20'; // There are two versions NASB1995 and NASB2020 on BLB
```
Comment (tobias-klein) [inline app/frontend/components/dictionary_info_box.js:264]: Thanks. This must be a recent change on BLB.

File: app/frontend/components/dictionary_info_box.js [hunk 3]
```diff
@@ -313,35 +311,27 @@ class DictionaryInfoBox {
...
-        var isLastRow = (i == (strongsEntry.references.length - 1));
-        var referenceTableRow = await this.getStrongsReferenceTableRow(strongsEntry.references[i], isLastRow);
-
-        if (referenceTableRow != null) {
-          extendedStrongsInfo += referenceTableRow;
-        }
-      }
+    var relatedStrongs = (await Promise.all(strongsEntry.references.map(async (ref, i) => 
+      await this.getStrongsReferenceTableRow(ref, i == (strongsEntry.references.length - 1))
+    ))).join('');
```
Comment (tobias-klein) [inline app/frontend/components/dictionary_info_box.js:316]: While this is a nice optimization, it's not so easy to ready any more. Would you mind adding a comment above this statement, that explains what's happening here?

It maybe even worth a separate function - then the function name would explain what is happening.

File: app/frontend/helpers/ui_helper.js [hunk 4]
```diff
@@ -93,86 +93,67 @@ class UiHelper {
...
-        verseListTabs.addClass('verse-list-tabs-tiny-screen');
-      } else {
-        verseListTabs.removeClass('verse-list-tabs-tiny-screen');
+      // Move dictionary to bottom panel
+      const dictionary = sidePanel.querySelector('#dictionary-info-box');
+      if (dictionary) {
+        bottomPanel.appendChild(dictionary);
+        bottomPanel.classList.add('with-dictionary');
+        sidePanel.classList.remove('with-dictionary');
+      }
```
Comment (tobias-klein) [inline app/frontend/helpers/ui_helper.js:140]: I suggest to move this to a separate function.

File: app/frontend/helpers/ui_helper.js [hunk 5]
```diff
@@ -93,86 +93,67 @@ class UiHelper {
...
+      sidePanel.style.display='';
+      app_controller.tag_assignment_menu.moveTagAssignmentList(false);
+
+      // Move dictionary back to side panel
+      const dictionary = bottomPanel.querySelector('#dictionary-info-box');
+      if (dictionary) {
+        sidePanel.appendChild(dictionary);
+        sidePanel.classList.add('with-dictionary');
+        bottomPanel.classList.remove('with-dictionary');
       }
```
Comment (tobias-klein) [inline app/frontend/helpers/ui_helper.js:152]: I suggest to move this to a separate function.

**Review by tobias-klein — DISMISSED** [review]
Thank you! This looks good.

Of course pre-condition for merging is that this also works on Android.
And would you mind documenting the grid layout with its areas in the DESIGN.md file?

**Comment by zhuiks** [comment]
> Of course pre-condition for merging is that this also works on Android.

I think I fixed it. 

But my fragile Android build system is broken :cry: after Android Studio and AVD update :cry: :cry:
I need some time to restore it. 

@tobias-klein Do you mind checking that it's working for Android. And if not, can you please provide screenshots.

File: DESIGN.md [hunk 6]
```diff
@@ -107,6 +107,21 @@ Here you find [i18next](https://www.i18next.com/) locale files (JSON format).
 
 Note that for adding a new locale you must also add an entry in the whitelist in `app/helpers/i18n_helper.js` (line 42).
 
+
+<a name='layout'></a>
+## App layout
```
Comment (tobias-klein) [inline DESIGN.md:112]: Would it make sense to include a picture? I think you once had some pictures in a discussion thread, right?

**Comment by tobias-klein** [comment]
> > Of course pre-condition for merging is that this also works on Android.
> 
> I think I fixed it.
> 
> But my fragile Android build system is broken 😢 after Android Studio and AVD update 😢 😢
> I need some time to restore it.
> 
> @tobias-klein Do you mind checking that it's working for Android. And if not, can you please provide screenshots.

I'll check how it looks on Android as soon as I can.

File: DESIGN.md (hunk 6 above)
Comment (zhuiks) [reply DESIGN.md:112]: Yes, including picture make sense

**Comment by tobias-klein** [comment]
> @tobias-klein Do you mind checking that it's working for Android. And if not, can you please provide screenshots.

I just tested this on an Android 10 device and the layout is working nicely. I am attaching screenshots anyway ... :)
I am going to run another test on an older Android device.

Screenshot_2021-08-07-09-10-57-122
Screenshot_2021-08-07-09-11-27-371

**Comment by tobias-klein** [comment]
This is how it looks on a Nexus 7 (Android 6).

Screenshot_20210807-103338
Screenshot_20210807-103348

**Comment by zhuiks** [comment]
@tobias-klein Thanks for the screenshots! I'm satisfied with the result!

Of course, the layout should be more optimized for smaller (7") screens. I'll try to do more prototyping and put it out for discussion under #299.
This grid layout should be a solid foundation for those future optimizations.

**Comment by zhuiks** [comment]
@tobias-klein can you confirm that #337 is fixed?

**Comment by tobias-klein** [comment]
Yes, I don't observe #337 anymore!

**Review by tobias-klein — APPROVED** [review]

**Pull request merged**